tests/cover/test_regex.py

```diff
@@ -0,0 +1,299 @@
+        c in UNICODE_SPACE_CHARS
+        for c in s
+    )
+
+
+def is_word(s):
+    return all(
+        c == '_' or (
+            (not HAS_WEIRD_WORD_CHARS or
+             c not in UNICODE_WEIRD_NONWORD_CHARS) and
+            unicodedata.category(c) in UNICODE_WORD_CATEGORIES
+        )
+        for c in s
+    )
+
+
+def ascii_regex(pattern):
+    flags = re.ASCII if PY3 else 0
+    return re.compile(pattern, flags)
+
+
+def unicode_regex(pattern):
+    return re.compile(pattern, re.UNICODE)
+
+
+def _test_matching_pattern(pattern, isvalidchar, is_unicode=False):
+    r = unicode_regex(pattern) if is_unicode else ascii_regex(pattern)
+
+    codepoints = hrange(0, sys.maxunicode + 1) \
+        if is_unicode else hrange(1, 128)
+    for c in [hunichr(x) for x in codepoints]:
+        if isvalidchar(c):
+            assert r.match(c), (
+                '"%s" supposed to match "%s" (%r, category "%s"), '
+                'but it doesnt' % (pattern, c, c, unicodedata.category(c))
+            )
+        else:
+            assert not r.match(c), (
+                '"%s" supposed not to match "%s" (%r, category "%s"), '
+                'but it does' % (pattern, c, c, unicodedata.category(c))
+            )
+
+
+@pytest.mark.parametrize('category,predicate', [
+    (r'\w', is_word), (r'\d', is_digit), (r'\s', None)])
+@pytest.mark.parametrize('invert', [False, True])
+@pytest.mark.parametrize('is_unicode', [False, True])
+def test_matching(category, predicate, invert, is_unicode):
+    if predicate is None:
+        # Special behaviour due to \x1c, INFORMATION SEPARATOR FOUR
+        predicate = is_unicode_space if is_unicode else is_space
+    pred = predicate
+    if invert:
+        category = category.swapcase()
+
+        def pred(s):
+            return not predicate(s)
+
+    _test_matching_pattern(category, pred, is_unicode)
+
+
+def assert_all_examples(strategy, predicate):
+    """Checks that there are no examples with given strategy that do not match
+    predicate.
+
+    :param strategy: Hypothesis strategy to check
+    :param predicate: (callable) Predicate that takes string example and
+        returns bool
+
+    """
+    @given(strategy)
+    def assert_examples(s):
+        assert predicate(s), \
+            'Found %r using strategy %s which does not match' % (s, strategy)
+
+    assert_examples()
+
+
+@pytest.mark.parametrize('pattern', [
+    u'.',  # anything
+    u'a', u'abc', u'[a][b][c]', u'[^a][^b][^c]',  # literals
+    u'[a-z0-9_]', u'[^a-z0-9_]',  # range and negative range
+    u'ab?', u'ab*', u'ab+',  # quantifiers
+    u'ab{5}', u'ab{5,10}', u'ab{,10}', u'ab{5,}',  # repeaters
+    u'ab|cd|ef',  # branch
+    u'(foo)+', u'([\'"])[a-z]+\\1',
+    u'(?:[a-z])([\'"])[a-z]+\\1', u'(?P<foo>[\'"])[a-z]+(?P=foo)',  # groups
+    u'^abc',  # beginning
+])
+def test_can_generate(pattern):
+    compiled_pattern = re.compile(pattern)
+    assert_all_examples(from_regex(pattern), compiled_pattern.match)
+    if PY3:
```

Comment:
Whoops - looks like I silenced this to let me work on other things and didn't reactivate them.  Plus of course there's a bug in my Py2 int-> bytes conversion...